core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/bi/BavetFilterBiConstraintStream.java

```diff
@@ -45,19 +47,63 @@ public boolean guaranteesDistinct() {
         return parent.guaranteesDistinct();
     }
 
+    // ************************************************************************
+    // Node creation
+    // ************************************************************************
+
+    @Override
+    public void collectActiveConstraintStreams(Set<BavetAbstractConstraintStream<Solution_>> constraintStreamSet) {
+        parent.collectActiveConstraintStreams(constraintStreamSet);
+        constraintStreamSet.add(this);
+    }
+
     @Override
-    public List<BavetFromUniConstraintStream<Solution_, Object>> getFromStreamList() {
-        return parent.getFromStreamList();
+    public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
+        Consumer<BiTuple<A, B>> insert = buildHelper.getAggregatedInsert(childStreamList);
+        Consumer<BiTuple<A, B>> retract = buildHelper.getAggregatedRetract(childStreamList);
+        buildHelper.putInsertRetract(this,
+                new ConditionalBiConsumer<>(predicate, insert),
+                retract);
+    }
+
+    private static final class ConditionalBiConsumer<A, B> implements Consumer<BiTuple<A, B>> {
+        private final BiPredicate<A, B> predicate;
+        private final Consumer<BiTuple<A, B>> consumer;
+
+        public ConditionalBiConsumer(BiPredicate<A, B> predicate, Consumer<BiTuple<A, B>> consumer) {
+            this.predicate = predicate;
+            this.consumer = consumer;
+        }
+
+        @Override
+        public void accept(BiTuple<A, B> tuple) {
+            if (predicate.test(tuple.factA, tuple.factB)) {
+                consumer.accept(tuple);
+            }
+        }
+
     }
 
     // ************************************************************************
-    // Node creation
+    // Equality for node sharing
     // ************************************************************************
 
     @Override
-    protected BavetFilterBiNode<A, B> createNode(BavetNodeBuildPolicy<Solution_> buildPolicy,
-            Score<?> constraintWeight, BavetAbstractBiNode<A, B> parentNode) {
-        return new BavetFilterBiNode<>(buildPolicy.getSession(), buildPolicy.nextNodeIndex(), parentNode, predicate);
+    public int hashCode() {
+        return Objects.hash(System.identityHashCode(parent), System.identityHashCode(predicate));
```

Comment:
Are you computing hashcodes of hashcodes?
I think the call to identity is not necessary here - the objects either are the same (= same hashcode) or they are not (=different hashcode).

Reply:
Good catch, this was nonsense.

Reply:
Fixed 3 cases in Bavet - but I did see some Score impl have some similar tactics: "Objects.hash(initScore, Arrays.hashCode(hardScores), Arrays.hashCode(softScores));" - I do think it might make sense there. Something to look into separately.